libraries/AP_DroneCAN/AP_DroneCAN_Filesystem.h

```diff
@@ -0,0 +1,125 @@
+#pragma once
+
+#include <uavcan.protocol.file.GetInfo.h>
+#include <uavcan.protocol.file.GetDirectoryEntryInfo.h>
+#include <uavcan.protocol.file.Delete.h>
+#include <uavcan.protocol.file.Read.h>
+#include <uavcan.protocol.file.Write.h>
+
+#include <AP_HAL/Semaphores.h>
+
+#if AP_FILESYSTEM_DRONECAN_ENABLED
+class AP_DroneCAN_Filesystem_Client
+{
+public:
+    AP_DroneCAN_Filesystem_Client(CanardInterface &canard_iface);
+    CLASS_NO_COPY(AP_DroneCAN_Filesystem_Client);
+
+    // Get info request and response
+    void request_info(const char* path, const uint8_t node_id);
+    bool request_info_response(uavcan_protocol_file_GetInfoResponse &msg);
+
+    // Get Directory Entry Info request and response
+    void request_directory_entry_info(const char* path, const uint8_t node_id, const uint32_t entry_index);
+    bool request_get_directory_entry_info_response(uavcan_protocol_file_GetDirectoryEntryInfoResponse &msg);
+
+    // Delete request and response
+    void request_delete(const char* path, const uint8_t node_id);
+    bool request_delete_response(uavcan_protocol_file_DeleteResponse &msg);
+
+    // Read request and response
+    void request_read(const char* path, const uint8_t node_id, const uint32_t offset);
+    bool request_read_response(uavcan_protocol_file_ReadResponse &msg);
+
+    // Write request and response
+    uint32_t request_write(const char* path, const uint8_t node_id, const uint32_t offset, const void *buf, uint32_t count);
+    bool request_write_response(uavcan_protocol_file_WriteResponse &msg);
+
+private:
+
+    // Get info request and response
+    Canard::Client<uavcan_protocol_file_GetInfoResponse> get_info_client;
+    Canard::ObjCallback<AP_DroneCAN_Filesystem_Client, uavcan_protocol_file_GetInfoResponse> get_info_cb{this, &AP_DroneCAN_Filesystem_Client::handle_get_info_response};
+    void handle_get_info_response(const CanardRxTransfer& transfer, const uavcan_protocol_file_GetInfoResponse &msg);
+    struct {
+        uavcan_protocol_file_GetInfoResponse msg;
+        bool valid;
+        HAL_Semaphore sem;
+    } get_info_response;
```

Comment:
I suspect this could be templated with some more effort. Only the `request.....` function is really different.